### PR TITLE
[MoE Layer] Add Grouped GEMM Fused Expert Weights Version

### PR DESCRIPTION
**Pull request by hushenwei2000**
### 简介

支持专家 weight 合并成一个大 weight 进行计算，节省 batched_gemm 计算前对所有专家权重聚合的 stack 操作。

支持单卡和 EP。

⚠️ 单卡情况下，本 PR 改为使用了 permute + unpermute 方式，导致性能下降严重，待修复。

loss 差异解释
1. 单卡情况下，目前非 grouped_gemm 版本是每个专家输出先乘以 weight，再合并，再 scatter；grouped_gemm 版本是先计算全部专家输出，先合并，再乘以 weight，再 scatter。
1. 单卡/多卡情况下，weight 初始化方式不同，目前使用手动尝试的方式找到了一个初始化方法和原来的精度类似。

### 使用方式

0. 原有 TransformerConfig：`moe_grouped_gemm` 默认关闭，需打开。若环境允许使用 deep_gemm 那么将使用，否则将用 Paddle batched_gemm。

**如果开启 Grouped GEMM 运行标准测试发生 OOM**，请配合 Paddle 的显存碎片整理使用。

1. 使用 Paddle Develop 版本

2. 在 PaddleFormers 的 trainer.py 增加以下一行，在优化器后整理显存
```
   self.optimizer_step(args, model=model, parameters_list=parameters_list)
+  paddle.device.cuda.vmm_compact()
```

3. 在执行训练之前开启以下四个 FLAGS
```
export FLAGS_use_virtual_memory_auto_growth=True  # 开启VMM机制
export FLAGS_vmm_small_pool_size_in_mb=1          # 开启大小池机制
export FLAGS_enable_compact_mem=True              # 开启碎片整理机制
export FLAGS_try_allocate=True                    # 开启TryAlloc（碎片整理检查更加严格）
```

#### 标准测试（acc 为 1）

本地测试多次 Paddle Grouped GEMM 与 deepgemm 性能持平，但 deepgemm 显存更优。

| | interval_tokens_per_second_per_device | 最大内存使用   | 最大内存预留 |
|--|---------------------------------------|----------|--------|
| for-loop                              | 6782.66  | 26.69  | 61.58 |
|                                       | 6614.03  | 26.69  | 61.58 |
|                                       | 6766.54  | 26.69  | 61.58 |
| paddle grouped_gemm                   | 10165.43 | 31.33  | 71.53 |
|                                       | 10403.42 | 31.33  | 71.53 |
|                                       | 10219.26 | 31.33  | 71.53 |
| deepgemm                              | 10337.53 | 26.72  | 62.09 |
|                                       | 10225.09 | 26.72  | 62.09 |
|                                       | 10314.53 | 26.72  | 62.09 |

#### 标准测试（acc 为 128）

Paddle Grouped GEMM 在标准测试下 OOM。

image

平均绝对误差: 0.002287
最大绝对误差: 0.014328

#### 模拟热启测试（专家权重全部设为常数值，acc 改为 1）

image

平均绝对误差: 0.000000

### 单测更新

grouped gemm 执行的 loss 值与原始 for-loop 版本间的差距如下，本 PR 不更新因为默认 moe_grouped_gemm 为开启。

GLM4.5 
```
Mismatched elements: 10 / 10 (100%)
Max absolute difference: 0.02793503
Max relative difference: 0.00231762
 x: array([12.080719, 12.025366, 12.041925, 12.040335, 12.012897, 11.988861,
       11.966151, 11.967965, 11.975366, 11.962843])
 y: array([12.104311, 12.053301, 12.038807, 12.034624, 12.020836, 12.008858,
       11.956372, 11.965539, 11.978788, 11.972153])
```

Qwen3
```
Mismatched elements: 10 / 10 (100%)
Max absolute difference: 0.00650406
Max relative difference: 0.00061803
 x: array([10.567345, 10.574219, 10.564425, 10.548807, 10.550143, 10.537522,
       10.530407, 10.523815, 10.540712, 10.52758 ])
 y: array([10.570881, 10.578835, 10.564507, 10.551708, 10.55018 , 10.537148,
       10.523903, 10.528399, 10.546388, 10.526874])
```

GLM4.5 集成测试
```
Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.01322651
Max relative difference: 0.00113954
 x: array([11.620065])
 y: array([11.606838])
```

### TODO

支持 checkpoint 保存/读取。
支持 deepgemm 在 batch_sizes 中的值不为 128 倍数时

### CODE REVIEW

File: src/paddlefleet/transformer/moe/fp8_utils.py [hunk 1]
```diff
@@ -19,6 +19,8 @@
 import paddle
 import paddle.nn.functional as F
 
+import paddlefleet
```
Comment (zhangbo9674) [inline src/paddlefleet/transformer/moe/fp8_utils.py]: fleet 内部需要 import fleet么

File: src/paddlefleet/transformer/moe/fp8_utils.py [hunk 2]
```diff
@@ -220,6 +222,8 @@ def __init__(
         backward_subbatch_rows=None,
         use_bf16_gemm_weight_grad=False,
         use_fp8_mlp=True,
+        moe_deep_gemm=True,
```
Comment (zhangbo9674) [inline src/paddlefleet/transformer/moe/fp8_utils.py:228]: megatron 也是叫这个么？

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 3]
```diff
@@ -18,13 +18,93 @@
 import paddle
 import paddle.nn.functional as F
 
+import paddlefleet
```
Comment (zhangbo9674) [inline src/paddlefleet/transformer/moe/moe_expert.py]: 同上

File: src/paddlefleet/transformer/moe/fp8_utils.py (hunk 2 above)
Comment (hushenwei2000) [reply src/paddlefleet/transformer/moe/fp8_utils.py:228]: Megatron 没有这个配置，目前因为 deepgemm 可能存在专家收到 token 数不对齐 128 的报错的情况，还是保留开关切换到框架 grouped gemm，后续修复了就不需要这个开关了

**Review by From00 — APPROVED** [review]
LGTM

**Review by Waynezee — APPROVED** [review]

**Review by swgu98 — APPROVED** [review]
